Sources/Algorithms/AdjacentPairs.swift

```diff
@@ -0,0 +1,271 @@
+///
+/// In an `AdjacentPairsCollection`, the elements of the *i*th pair are the *i*th
+/// and *(i+1)*th elements of the underlying sequence. The following example
+/// uses the `adjacentPairs()` method to iterate over adjacent pairs of
+/// integers:
+/// ```
+/// for pair in (1...5).adjacentPairs() {
+///     print(pair)
+/// }
+/// // Prints "(1, 2)"
+/// // Prints "(2, 3)"
+/// // Prints "(3, 4)"
+/// // Prints "(4, 5)"
+/// ```
+public struct AdjacentPairsCollection<Base: Collection> {
+  @usableFromInline
+  internal let base: Base
+
+  public let startIndex: Index
+
+  @inlinable
+  internal init(base: Base) {
+    self.base = base
+
+    // Precompute `startIndex` to ensure O(1) behavior,
+    // avoiding indexing past `endIndex`
+    let start = base.startIndex
+    let end = base.endIndex
+    let second = start == end ? start : base.index(after: start)
+    self.startIndex = Index(first: start, second: second)
+  }
+}
+
+extension AdjacentPairsCollection {
+  public typealias Iterator = AdjacentPairsSequence<Base>.Iterator
+
+  @inlinable
+  public func makeIterator() -> Iterator {
+    Iterator(base: base.makeIterator())
+  }
+}
+
+extension AdjacentPairsCollection {
+  public struct Index: Comparable {
+    @usableFromInline
+    internal var first: Base.Index
+    
+    @usableFromInline
+    internal var second: Base.Index
+
+    @inlinable
+    internal init(first: Base.Index, second: Base.Index) {
+      self.first = first
+      self.second = second
+    }
+
+    @inlinable
+    public static func < (lhs: Index, rhs: Index) -> Bool {
+      (lhs.first, lhs.second) < (rhs.first, rhs.second)
+    }
+  }
+}
+
+extension AdjacentPairsCollection: Collection {
+  @inlinable
+  public var endIndex: Index {
+    switch base.endIndex {
+    case startIndex.first, startIndex.second:
+      return startIndex
+    case let end:
+      return Index(first: end, second: end)
+    }
+  }
+
+  @inlinable
+  public subscript(position: Index) -> (Base.Element, Base.Element) {
+    (base[position.first], base[position.second])
+  }
+
+  @inlinable
+  public func index(after i: Index) -> Index {
+    let next = base.index(after: i.second)
+    return next == base.endIndex
+      ? endIndex
+      : Index(first: i.second, second: next)
+  }
+
+  @inlinable
+  public func index(_ i: Index, offsetBy distance: Int) -> Index {
+    if distance == 0 {
+      return i
+    } else if distance > 0 {
+      let firstOffsetIndex = base.index(i.first, offsetBy: distance)
+      let secondOffsetIndex = base.index(after: firstOffsetIndex)
+      return secondOffsetIndex == base.endIndex
+        ? endIndex
+        : Index(first: firstOffsetIndex, second: secondOffsetIndex)
+    } else {
+      return i == endIndex
+        ? Index(first: base.index(i.first, offsetBy: distance - 1),
+                second: base.index(i.first, offsetBy: distance))
+        : Index(first: base.index(i.first, offsetBy: distance),
+                second: i.first)
```

Comment:
You could avoid computing the first and second indices separately here, since you can cheaply compute the firts if you already have the second.